asserts/snapasserts/validation_sets.go

```diff
@@ -0,0 +1,244 @@
+		}
+		msg += fmt.Sprintf(" at different revisions %s", strings.Join(l, ", "))
+	}
+
+	if unspecifiedOnes, ok := c.revisions[unspecifiedRevision]; ok {
+		which := whichSets(unspecifiedOnes, asserts.PresenceRequired)
+		if which != "" {
+			if len(revnos) != 0 {
+				msg += " or"
+			}
+			if invalid {
+				msg += fmt.Sprintf(" at any revision %s", which)
+			} else {
+				msg += fmt.Sprintf(" required at any revision %s", which)
+			}
+		}
+	}
+	return conflicting, fmt.Errorf(msg)
+}
+
+// NewValidationSets returns a new ValidationSets.
+func NewValidationSets() *ValidationSets {
+	return &ValidationSets{
+		sets:  map[string]*asserts.ValidationSet{},
+		snaps: map[string]*snapContraints{},
+	}
+}
+
+func valSetKey(valset *asserts.ValidationSet) string {
+	return fmt.Sprintf("%s/%s", valset.AccountID(), valset.Name())
+}
+
+// Add adds the given asserts.ValidationSet to the combination.
+// It errors if a validation-set with the same sequence key has been
+// added already.
+func (v *ValidationSets) Add(valset *asserts.ValidationSet) error {
+	k := valSetKey(valset)
+	if _, ok := v.sets[k]; ok {
+		return fmt.Errorf("cannot add a second validation-set under %q", k)
+	}
+	v.sets[k] = valset
+	for _, sn := range valset.Snaps() {
+		v.addSnap(sn, k)
+	}
+	return nil
+}
+
+func (v *ValidationSets) addSnap(sn *asserts.ValidationSetSnap, validationSetKey string) {
+	rev := snap.R(sn.Revision)
+	if sn.Presence == asserts.PresenceInvalid {
+		rev = invalidPresRevision
+	}
+
+	rc := &revConstraint{
+		validationSetKey:  validationSetKey,
+		ValidationSetSnap: *sn,
+	}
+
+	cs := v.snaps[sn.SnapID]
+	if cs == nil {
+		v.snaps[sn.SnapID] = &snapContraints{
+			name:     sn.Name,
+			presence: sn.Presence,
+			revisions: map[snap.Revision][]*revConstraint{
+				rev: {rc},
+			},
+		}
+		return
+	}
+
+	cs.revisions[rev] = append(cs.revisions[rev], rc)
+	if cs.presence == presConflict {
+		// nothing to check anymore
+		return
+	}
+	// this counts really different revisions or invalid
+	ndiff := len(cs.revisions)
+	if _, ok := cs.revisions[unspecifiedRevision]; ok {
+		ndiff -= 1
+	}
+	switch {
+	case cs.presence == asserts.PresenceOptional:
+		cs.presence = sn.Presence
+		fallthrough
+	case cs.presence == sn.Presence || sn.Presence == asserts.PresenceOptional:
+		if ndiff > 1 {
+			if cs.presence == asserts.PresenceRequired {
+				// different revisions required/invalid
+				cs.presence = presConflict
+				return
+			}
+			// multiple optional at different revisions => invalid
+			cs.presence = asserts.PresenceInvalid
+		}
+		return
+	}
+	// we are left with a combo of required and invalid => conflict
+	cs.presence = presConflict
+	return
+}
+
+// Conflict returns a non-nil error if the combination is in conflict,
+// nil otherwise.
+func (v *ValidationSets) Conflict() error {
+	sets := make(map[string]*asserts.ValidationSet)
+	snaps := make(map[string]error)
+
+	for snapID, snConstrs := range v.snaps {
+		conflicting, err := snConstrs.conflict()
```

Comment:
If you take my suggestion then this side needs to change as well, and will most likely read more expectedly.